### PR TITLE
MULE-9732 DefaultExtensionFactory is setting the incorrect ClassLoader

### PR DESCRIPTION
**Pull request by gsfernandes**
_Fixes issue when factory creates the extension model that was using the incorrect classloader
_Also fixes a similar issue when metadata was exposed for an extension, it was not using the extension classloader

### CODE REVIEW

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/ExtensionComponent.java [hunk 1]
```diff
@@ -138,7 +141,8 @@ public void setMuleContext(MuleContext context)
     @Override
     public MetadataResult<ComponentMetadataDescriptor> getMetadata(MetadataKey key) throws MetadataResolvingException
     {
-        return metadataMediator.getMetadata(getMetadataContext(), key);
+        final MetadataContext metadataContext = getMetadataContext();
+        return withContextClassLoader(MuleExtensionUtils.getClassLoader(this.extensionModel), () -> metadataMediator.getMetadata(metadataContext, key));
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/ExtensionComponent.java]: static import

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/processor/OperationMessageProcessor.java [hunk 2]
```diff
@@ -75,7 +76,7 @@ public OperationMessageProcessor(RuntimeExtensionModel extensionModel,
                                      ResolverSet resolverSet,
                                      ExtensionManagerAdapter extensionManager)
     {
-        super(extensionModel, operationModel, configurationProviderName, extensionManager);
+        super(extensionModel, configurationProviderName, extensionManager, new MetadataMediator(extensionModel, operationModel));
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/processor/OperationMessageProcessor.java]: why this change?

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/ExtensionComponent.java [hunk 3]
```diff
@@ -65,12 +67,12 @@
...
     private MuleMetadataManager metadataManager;
 
-    protected ExtensionComponent(RuntimeExtensionModel extensionModel, RuntimeComponentModel componentModel, String configurationProviderName, ExtensionManagerAdapter extensionManager)
+    protected ExtensionComponent(RuntimeExtensionModel extensionModel, String configurationProviderName, ExtensionManagerAdapter extensionManager, MetadataMediator metadataMediator)
     {
         this.extensionModel = extensionModel;
         this.configurationProviderName = configurationProviderName;
         this.extensionManager = extensionManager;
-        this.metadataMediator = new MetadataMediator(extensionModel, componentModel);
+        this.metadataMediator = metadataMediator;
```
Comment (marianogonzalez) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/ExtensionComponent.java]: why should the Mediator be exposed like this?

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/source/ExtensionMessageSourceTestCase.java [hunk 4]
```diff
@@ -169,6 +186,64 @@ public void before() throws Exception
     }
 
     @Test
+    public void getMetadataUsesClassLoaderModelProperty() throws MetadataResolvingException
+    {
+        final MetadataResult result = mock(MetadataResult.class);
```
Comment (marianogonzalez) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/source/ExtensionMessageSourceTestCase.java]: this is testing the implementation not the behaviour. If what you want is to make sure that the right classloader is used, then you can just use a custom Answer object that verifies the TCCL

File: modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/ExtensionComponent.java (hunk 3 above)
Comment (gsfernandes) [inline modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/runtime/ExtensionComponent.java]: Actually I was about to make an interface first and them added like this as constructor parameter in order to be able to mock it because when I wanted to test the ExtensionComponent classloader changes I was not able to do it without mocking this class.

What do you think? Should we expose this mediator object to be injected so it can be "changed" later (on this case, mocked)

File: modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/source/ExtensionMessageSourceTestCase.java (hunk 4 above)
Comment (gsfernandes) [inline modules/extensions-support/src/test/java/org/mule/runtime/module/extension/internal/runtime/source/ExtensionMessageSourceTestCase.java]: I didn't get this. What I want to test is the following code in ExtensionComponent:
```
    @Override
    public MetadataResult<ComponentMetadataDescriptor> getMetadata() throws MetadataResolvingException
    {
        return withContextClassLoader(getClassLoader(this.extensionModel), () -> metadataMediator.getMetadata());
    }
```
When mediator is called the correct TCCL should be set. In order to test that I did use a custom Answer but at the mediator.getMetadata() mocked method. 

I didn't like it how the has to be written but didn't know another way.
Is there a better way to test this?

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java [hunk 5]
```diff
@@ -511,4 +533,31 @@ public void interfaceWithSubtypesMetadataType()
...
+     */
+    private void doTestThreadContextClassLoaderResolver(String flowName, Callback<OperationMetadataTestCase> doAction) throws Exception
+    {
+        componentId = new ProcessorId(flowName, FIRST_PROCESSOR_INDEX);
+        TestThreadContextClassLoaderResolver.reset();
+        final ClassLoader mockClassLoader = mock(ClassLoader.class);
+        final ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
+        try
+        {
+            Thread.currentThread().setContextClassLoader(mockClassLoader);
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java]: ClassUtils has a withContextClassLoader method which does this

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java [hunk 6]
```diff
@@ -393,6 +397,24 @@ public void dynamicOutputAndContentWithCache() throws Exception
     }
 
     @Test
+    public void threadContextClassLoaderResolverTypeKeys() throws Exception
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java]: rename to resolveKeysWithClassLoader

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java [hunk 7]
```diff
@@ -393,6 +397,24 @@ public void dynamicOutputAndContentWithCache() throws Exception
...
+    {
+        doTestThreadContextClassLoaderResolver(THREAD_CONTEXT_CLASSLOADER_RESOLVER_KEYS, source -> source.metadataManager.getMetadataKeys(componentId));
+    }
+
+    @Test
+    public void threadContextClassLoaderResolverContent() throws Exception
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java]: rename to resolveContentWithClassLoader

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java [hunk 8]
```diff
@@ -393,6 +397,24 @@ public void dynamicOutputAndContentWithCache() throws Exception
...
+    {
+        doTestThreadContextClassLoaderResolver(THREAD_CONTEXT_CLASSLOADER_RESOLVER_CONTENT, source -> source.getComponentDynamicMetadata());
+    }
+
+    @Test
+    public void threadContextClassLoaderResolverOutput() throws Exception
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java]: idem

File: modules/extensions-spring-support/src/test/resources/metadata-tests.xml [hunk 9]
```diff
@@ -89,6 +89,18 @@
         <metadata:content-only-cache-resolver config-ref="config" type="Person"/>
     </flow>
 
+    <flow name="threadContextClassLoaderResolverTypeKeys">
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/resources/metadata-tests.xml]: improve flow names

File: tests/test-extensions/metadata-extension/src/main/java/org/mule/test/metadata/extension/MetadataOperations.java [hunk 10]
```diff
@@ -131,6 +132,22 @@ public MuleMessage messageAttributesNullTypeMetadata()
         return null;
     }
 
+    @MetadataScope(keysResolver = TestThreadContextClassLoaderResolver.class)
+    public void threadContextClassLoaderResolverTypeKeys(@MetadataKeyId String type)
```
Comment (marianogonzalez) [inline tests/test-extensions/metadata-extension/src/main/java/org/mule/test/metadata/extension/MetadataOperations.java]: better names

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java [hunk 11]
```diff
@@ -511,4 +534,32 @@ public void interfaceWithSubtypesMetadataType()
...
+    {
+        componentId = new ProcessorId(flowName, FIRST_PROCESSOR_INDEX);
+        TestThreadContextClassLoaderResolver.reset();
+        final ClassLoader mockClassLoader = mock(ClassLoader.class);
+        final ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
+        try
+        {
+            ClassUtils.withContextClassLoader(mockClassLoader, () -> {
+                doAction.execute(OperationMetadataTestCase.this);
+                assertThat(TestThreadContextClassLoaderResolver.getCurrentState(), is(sameInstance(originalClassLoader)));
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java]: shouldn't the mockClassLoader be used here?

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java [hunk 12]
```diff
@@ -511,4 +534,32 @@ public void interfaceWithSubtypesMetadataType()
...
+     */
+    private void doResolverTestWithContextClassLoader(String flowName, Callback<OperationMetadataTestCase> doAction) throws Exception
+    {
+        componentId = new ProcessorId(flowName, FIRST_PROCESSOR_INDEX);
+        TestThreadContextClassLoaderResolver.reset();
+        final ClassLoader mockClassLoader = mock(ClassLoader.class);
+        final ClassLoader originalClassLoader = Thread.currentThread().getContextClassLoader();
+        try
+        {
+            ClassUtils.withContextClassLoader(mockClassLoader, () -> {
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java]: static import

File: modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java [hunk 13]
```diff
@@ -511,4 +534,32 @@ public void interfaceWithSubtypesMetadataType()
...
+                doAction.execute(OperationMetadataTestCase.this);
+                assertThat(TestThreadContextClassLoaderResolver.getCurrentState(), is(sameInstance(originalClassLoader)));
+            });
+        }
+        finally
+        {
+            Thread.currentThread().setContextClassLoader(originalClassLoader);
```
Comment (marianogonzalez) [inline modules/extensions-spring-support/src/test/java/org/mule/runtime/module/extension/internal/metadata/OperationMetadataTestCase.java]: you don't need this anymore. withContextClassLoader does that automatically
